### PR TITLE
Fix reciprocal tests for zero values and improve test value customization

### PR DESCRIPTION
**Pull request by jjwilke**
Reciprocal is not valid for integers, which leads to test failures on certain platforms. 

https://numpy.org/doc/stable/reference/generated/numpy.reciprocal.html

* Splits math ufunc tests into default operations and operations needing special values
* Enables easier customization of input values to the different tests
* Simplifies syntax for calling tests since all tests only take a single argument
* Uses a hash-based seed for initializing random values so that tests use the same values whether tests are run individually or all-at-once

### CODE REVIEW

**Comment by bryevdv** [comment]
@jjwilke please see https://github.com/nv-legate/cunumeric/pull/457 which splits all the ufuncs into parameterized groups. Could you wait to do this work once that is merged?

**Comment by jjwilke** [comment]
> @jjwilke please see #457 which splits all the ufuncs tests into parameterized groups. Could you wait to do this work once that is merged? (I've already had to fix three merge conflicts and it's not even 10 am 😃 I think it will be simpler for you to update this based on #457, than vice-versa)

Ah, nice. Good thing I didn't follow through on my original plan to also parameterize all the tests : ) Yep, I can rebase once #457 is merged.

**Comment by magnatelee** [comment]
@jjwilke I should have read the API reference more carefully when filling in these ufuncs! FYI, I've been referring to type signatures of NumPy's ufuncs when deciding for which types cuNumeric's operator templates should be instantiated; they nominally mean there would be specialized implementations for them (as opposed to those that can be derived from one with the closest signature and type conversions on the operands). Interestingly, type signatures of `numpy.reciprocal` contain integer types, at least on Linux:
```
>>> np.reciprocal.types
['b->b', 'B->B', 'h->h', 'H->H', 'i->i', 'I->I', 'l->l', 'L->L', 'q->q', 'Q->Q', 'e->e', 'f->f', 'd->d', 'g->g', 'F->F', 'D->D', 'G->G', 'O->O']
``` 
So I had to instantiate `cunumeric.reciprocal` for integer types as well and give a reasonable semantics to the case when the operand is 0. I wouldn't realize `numpy.reciprocal` wasn't actually well defined for integer types.

I skimmed through the ufuncs just to see if there's any other ufunc like `numpy.reciprocal` (i.e., not well-defined on integers), but I couldn't find any, so this PR should be a sufficient fix.

**Comment by jjwilke** [comment]
@magnatelee Huh, good call. I didn't think to check the dtypes.  Looking at it more closely, the only difference is how `numpy` treats `1/0`. cunumeric returns `0`, but numpy returns `-1`.

This is all obviously pretty low priority (other than making sure tests pass). Another option would be to test integer types since these are valid types, but ensure that all values are non-zero.

**Comment by magnatelee** [comment]
I prefer making sure the integer values are non-zero to not testing them at all.

**Comment by bryevdv** [comment]
@jjwilke FYI #457 has just been merged

**Comment by jjwilke** [comment]
@magnatelee @bryevdv This should be ready for re-review

File: tests/integration/test_unary_ufunc.py [hunk 1]
```diff
@@ -189,27 +253,27 @@ def test_trig_ops(op):
 
 
 @pytest.mark.parametrize("op", arc_hyp_trig_ops)
-def test_arc_hyp_trig_ops(op):
-    check_op(op, (np.random.uniform(low=1, high=5, size=(4, 5)),))
-    check_op(op, (np.random.uniform(low=1, high=5, size=(4, 5)).astype("e"),))
-    check_op(op, (np.array(np.random.uniform(low=1, high=5)),))
+def test_arc_hyp_rig_ops(op):
```
Comment (bryevdv) [inline tests/integration/test_unary_ufunc.py:256]:
```suggestion
def test_arc_hyp_trig_ops(op):
```

**Review by bryevdv — APPROVED** [review]

**Comment by bryevdv** [comment]
LGTM

**Comment by magnatelee** [comment]
@jjwilke some GPU tests seem to be failing

**Comment by jjwilke** [comment]
So we deterministically set a random seed for the testing. But that means if an extra test gets added or reordered, this will change the inputs to other tests. This appears to be what happened to the `test_trig_ops`. Nothing functionally changed, but the inputs must have changed.

The differences observed were very small and they only appeared for fp16. Our current rtol = 1e-5, but fp16 is only good for rtol = 1e-4. I've modified this to lower the `isclose` precision if it detects half-precision.

**Comment by manopapad** [comment]
On contraction tests that involve half-precision (https://github.com/nv-legate/cunumeric/blob/branch-22.07/tests/integration/utils/contractions.py#L117-L122) I've had to adjust tolerance even lower, to `rtol=0.01`, presumably due to error accumulation.

**Comment by jjwilke** [comment]
@magnatelee Ready for re-review

**Review by magnatelee — APPROVED** [review]

**Comment by magnatelee** [comment]
looks good to me, but please address this comment from @bryevdv: https://github.com/nv-legate/cunumeric/pull/467/files#r931659174

**Review by bryevdv — APPROVED** [review]
